contrib/design-docs/config-file-parsing.md

```diff
@@ -0,0 +1,342 @@
+field_4 = user
+field_6 = f
+```
+
+Now parsing this as user with UID 1000 results in this final config:
+
+```
+field_2 = b
+field_4 = d
+field_5 = e
+field_6 = f
+```
+
+#### Environment Variables
+
+The following two envs should be defined for each config file:
+
+`CONTAINERS_<name>_CONF`: If set only read the given file in this env and nothing else.
+`CONTAINERS_<name>_CONF_OVERRIDE`: If set append the given file as last file after parsing
+all other files as normal. Useful to overwrite a single field for testing without overwriting
+the rest of the system configuration.
+
+As special case for containers.conf the name of the vars is `CONTAINERS_CONF` and `CONTAINERS_CONF_OVERRIDE`.
+
+The handling of these should be part of the `configfile` package.
+
+#### Appending arrays
+
+The toml parser by default replaces arrays in each file which makes it impossible to append values in drop-ins, etc...
+
+containers.conf already has a work for that with a custom syntax to trigger appending:
+```
+field = ["val", {append=true}]
+```
+I propose we adapt the same universally for the other config files.
+
+https://github.com/containers/container-libs/blob/main/common/docs/containers.conf.5.md#appending-to-string-arrays
+
+This means moving the `common/internal/attributedstring` into the new configfile package so all callers can use it.
+
+#### Scope
+
+##### containers.conf
+
+No changes except `/etc/containers/containers.rootless.conf` search location has been removed. It has just been added
+in 5.7 so I don't think it would cause major concerns to drop it again.
+
+The reason to not support the main file with rootless/rootful now is that it is not obvious how this should interact
+with the parsing, should it replace the main config file or act like a drop in?  As such I think it is better to not
+support this and we should generally push all users to use a drop instead of editing the main file.
+
+Also there was/is some discussion of splitting containers.conf in two files as currently there are fields in there
+that are only read on the server side while others only get used on the client side which makes using it in a remote
+context such as podman machine confusing. For now this is not part of this design doc, we may make another design docs
+just for this in case we like to move forward on it.
+
+containers.conf also supports "Modules", i.e. `podman --module name.conf ...` which adds additional drop-in files at
+the end after the regular config files. This functionality should be preserved but don't expand module support to the
+other files.
+
+##### storage.conf
+
+Deprecate `rootless_storage_path` option. With the `rootless/rootful` config location and admin could just use
+`graphroot` in the location in the rootless file location. As such there is no need to special case these fields
+in the parser.
+
+String arrays in the config will need to get switched to the attributedstring type, as described under Appending arrays.
+
+##### registries.conf
```

Comment:
There’s another different behavior, see `newConfigWrapperWithHomeDir` : If there is a per-user `registries.conf`, the system-wide `registries.conf.d` are completely ignored. That was designed that way intentionally, historically; the UAPI spec explicitly defines the opposite.

I don’t know what we want to do here…